### ISSUE TITLE
slicing bug

### ISSUE THREAD

**Issue by crusaderky** [dask/dask#1770]
```
>> d1 = dask.array.zeros((1, 9, 3), chunks=(1, 5, 3))
>> d2 = d1[0, :, [0, 2]]
>> d2.compute()
ValueError: could not broadcast input array from shape (2,4) into shape (2,5)
```
I tried looking at the dask dict but I could not notice anything anomalous (besides the fact that the slicing is done in two steps and may have been a bit more optimized).

I tried a bunch of variants of the above:
- do not slice on the first dim: WORKS
- do not have a first dim: WORKS
- slice on the first dim alone, compute to numpy, convert back to dask, and then slice on the third dim: WORKS
- select only one element on the third dim: WORKS
- have 10 elements in the second dim, so that it can be divided evenly by its chunksize: WORKS
- slice on the first dim, and then slice the result on the third dim: DOESN'T WORK

Expanded tests:
https://gist.github.com/crusaderky/87e6e51b22434128a3a57811a4077710

**Comment by mrocklin** [comment]
Thanks for the excellent bug report.  Reproduced on my side.

**Comment by laliberte** [comment]
I've done some investigating this bug and it seems to be a basic feature of [combined advanced and basic slicing](https://docs.scipy.org/doc/numpy/reference/arrays.indexing.html#combining-advanced-and-basic-indexing) in numpy. 

In short, in this example `d1[0, :, [0, 2])` falls in the first case (as explained in the link above), i.e. the "advanced" slices are separated by basic slicing (the Ellipsis). In that specific case, the advanced slicing dimension (the third dimension) is then moved to the first dimension in the output:
```
>> a = np.arange(27).reshape((1, 9, 3))
>> a[0, :, [0, 2]].shape
(2, 9)
>> np.squeeze(a[:, :, [0,2]]).shape
(9, 2)
```
and `a[0, :, [0, 2]].T` is going to be equal to `np.squeeze(a[:, :, [0, 2]])`.

My understanding is that this problem arises in the optimization step where `a[0, :, :][:, [0, 2]]` gets converted to `a[0, :, [0, 2]]`. This then reverts to `numpy` indexing rules yielding an unexpexted array shape to dask.

Because of this, I think there is only one clean option: do not optimize these special cases...

**Comment by jcrist** [comment]
The root bug was fixed in #2313. Specific optimizations for hdf5 were not part of that though. I'm going to close this, but keep PR #1776 open for discussion.
